File: go/envelope_test.go
package plutus

import (
	"encoding/base64"
	"errors"
	"strings"
	"testing"
)

func TestEncryptedEnvelopeVectors(t *testing.T) {
	for _, vec := range vectors(t).Vectors.EncryptedEnvelope {
		vec := vec
		t.Run(vec.ID, func(t *testing.T) {
			aad := AAD{
				RequestID:     vec.AADComponents.RequestID,
				RouteTemplate: vec.AADComponents.RouteTemplate,
				Timestamp:     vec.AADComponents.Timestamp,
				KeyID:         vec.AADComponents.KeyID,
			}
			if aad.String() != vec.AADString {
				t.Fatalf("aad string = %q, want %q", aad.String(), vec.AADString)
			}
			if aad.Base64() != vec.AADBase64 {
				t.Fatalf("aad base64 = %q, want %q", aad.Base64(), vec.AADBase64)
			}
			// 信封里的 aad 只是回显, 必须与本地重建值一致。
			if vec.Envelope.AAD != vec.AADBase64 {
				t.Fatalf("envelope aad echo = %q, want %q", vec.Envelope.AAD, vec.AADBase64)
			}
			if vec.Envelope.Algorithm != EnvelopeAlgorithm {
				t.Fatalf("algorithm = %q, want %q", vec.Envelope.Algorithm, EnvelopeAlgorithm)
			}
			if vec.Envelope.EncryptedPayload != vec.Envelope.Ciphertext {
				t.Fatalf("encryptedPayload must equal ciphertext on the API chain")
			}
			if vec.Envelope.EnvelopeVersion != nil {
				t.Fatalf("API chain envelope must not carry envelopeVersion")
			}
			if got, err := vec.Envelope.DecodedAAD(); err != nil || got != aad {
				t.Fatalf("DecodedAAD = %+v (%v), want %+v", got, err, aad)
			}

			priv := privateKey(t, vec.DecryptionKey)
			if fp, err := KeyFingerprint(&priv.PublicKey); err != nil || fp != vec.Envelope.KeyFingerprint {
				t.Fatalf("keyFingerprint = %q, want decryption key fingerprint %q (%v)", vec.Envelope.KeyFingerprint, fp, err)
			}

			envelope := vec.Envelope
			plaintext, err := Open(priv, &envelope, aad)
			if err != nil {
				t.Fatalf("Open: %v", err)
			}
			if string(plaintext) != vec.ExpectedPlaintext {
				t.Fatalf("plaintext = %q, want %q", plaintext, vec.ExpectedPlaintext)
			}

			// 敏感响应方向: 高层入口从回显 aad 取 timestamp 并重建其余分量。
			if vec.Direction == "sensitive_response" {
				got, err := OpenSensitiveResponse(priv, &envelope, aad.RequestID)
				if err != nil {
					t.Fatalf("OpenSensitiveResponse: %v", err)
				}
				if string(got) != vec.ExpectedPlaintext {
					t.Fatalf("sensitive plaintext = %q, want %q", got, vec.ExpectedPlaintext)
				}
				// 关联 ID 不一致时 AAD 重建失败。
				if _, err := OpenSensitiveResponse(priv, &envelope, aad.RequestID+"x"); !errors.Is(err, ErrAADMismatch) {
					t.Errorf("wrong request id: got %v, want ErrAADMismatch", err)
				}
			}

			// 负向: AAD 任一分量被篡改都必须失败。
			for name, tampered := range map[string]AAD{
				"requestId":     {RequestID: aad.RequestID + "x", RouteTemplate: aad.RouteTemplate, Timestamp: aad.Timestamp, KeyID: aad.KeyID},
				"routeTemplate": {RequestID: aad.RequestID, RouteTemplate: aad.RouteTemplate + "/x", Timestamp: aad.Timestamp, KeyID: aad.KeyID},
				"timestamp":     {RequestID: aad.RequestID, RouteTemplate: aad.RouteTemplate, Timestamp: "1755600099000", KeyID: aad.KeyID},
				"keyId":         {RequestID: aad.RequestID, RouteTemplate: aad.RouteTemplate, Timestamp: aad.Timestamp, KeyID: "SHA256:" + strings.Repeat("0", 64)},
			} {
				if _, err := Open(priv, &envelope, tampered); !errors.Is(err, ErrAADMismatch) {
					t.Errorf("tampered aad component %s: got %v, want ErrAADMismatch", name, err)
				}
			}

			// 负向: 篡改密文时 GCM 认证必须失败。
			blob, err := base64.StdEncoding.DecodeString(envelope.Ciphertext)
			if err != nil {
				t.Fatalf("decode ciphertext: %v", err)
			}
			blob[len(blob)-1] ^= 0x01
			corrupted := envelope
			corrupted.Ciphertext = base64.StdEncoding.EncodeToString(blob)
			if _, err := Open(priv, &corrupted, aad); !errors.Is(err, ErrDecryptFailed) {
				t.Errorf("tampered ciphertext: got %v, want ErrDecryptFailed", err)
			}

			// 负向: 用另一把私钥解包必须失败。
			other := privateKey(t, "platform_auth")
			if _, err := Open(other, &envelope, aad); !errors.Is(err, ErrDecryptFailed) {
				t.Errorf("wrong private key: got %v, want ErrDecryptFailed", err)
			}
		})
	}
}

func TestSealRequestRoundTrip(t *testing.T) {
	pub := publicKey(t, "platform_enc")
	priv := privateKey(t, "platform_enc")
	fingerprint, err := KeyFingerprint(pub)
	if err != nil {
		t.Fatalf("fingerprint: %v", err)
	}

	const (
		requestID     = "req_roundtrip_0001"
		routeTemplate = "/card-products/10010106/shared/cards/create"
		timestamp     = "1755600010000"
	)
	plaintext := []byte(`{"platformCardProductBizId":"pcp_example_001","quantity":2}`)

	env, err := SealRequest(pub, requestID, routeTemplate, timestamp, plaintext)
	if err != nil {
		t.Fatalf("SealRequest: %v", err)
	}
	if env.KeyFingerprint != fingerprint {
		t.Errorf("keyFingerprint = %q, want %q", env.KeyFingerprint, fingerprint)
	}
	if env.EncryptedPayload != env.Ciphertext {
		t.Error("encryptedPayload must equal ciphertext")
	}
	if env.EnvelopeVersion != nil {
		t.Error("API chain envelope must not carry envelopeVersion")
	}
	aad := AAD{RequestID: requestID, RouteTemplate: routeTemplate, Timestamp: timestamp, KeyID: fingerprint}
	if env.AAD != aad.Base64() {
		t.Errorf("aad echo = %q, want %q", env.AAD, aad.Base64())
	}

	blob, err := base64.StdEncoding.DecodeString(env.Ciphertext)
	if err != nil {
		t.Fatalf("decode ciphertext: %v", err)
	}
	if want := GCMNonceSize + len(plaintext) + GCMTagSize; len(blob) != want {
		t.Errorf("ciphertext blob = %d bytes, want %d (iv||ct||tag)", len(blob), want)
	}

	got, err := Open(priv, env, aad)
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	if string(got) != string(plaintext) {
		t.Fatalf("round-trip plaintext = %q, want %q", got, plaintext)
	}

	// 每次加密使用新的一次性密钥与 IV。
	again, err := SealRequest(pub, requestID, routeTemplate, timestamp, plaintext)
	if err != nil {
		t.Fatalf("SealRequest: %v", err)
	}
	if again.Ciphertext == env.Ciphertext || again.EncryptedKey == env.EncryptedKey {
		t.Error("envelope must use a fresh AES key and IV on every seal")
	}

	if _, err := SealRequest(pub, "", routeTemplate, timestamp, plaintext); !errors.Is(err, ErrInvalidConfig) {
		t.Error("empty request id must be rejected")
	}
	if _, err := SealRequest(pub, requestID, routeTemplate, "1755600010", plaintext); !errors.Is(err, ErrInvalidTimestamp) {
		t.Error("second-precision timestamp must be rejected")
	}
}

func TestSensitiveResponseRoundTrip(t *testing.T) {
	pub := publicKey(t, "merchant_enc")
	priv := privateKey(t, "merchant_enc")
	fingerprint, err := KeyFingerprint(pub)
	if err != nil {
		t.Fatalf("fingerprint: %v", err)
	}
	aad := AAD{RequestID: "req_sensitive_0001", RouteTemplate: "", Timestamp: "1755600011000", KeyID: fingerprint}
	env, err := Seal(pub, fingerprint, aad, []byte(`{"cardNumber":"4111111111111111"}`))
	if err != nil {
		t.Fatalf("Seal: %v", err)
	}
	plaintext, err := OpenSensitiveResponse(priv, env, aad.RequestID)
	if err != nil {
		t.Fatalf("OpenSensitiveResponse: %v", err)
	}
	if string(plaintext) != `{"cardNumber":"4111111111111111"}` {
		t.Fatalf("plaintext = %q", plaintext)
	}

	// 指纹不符 (平台用了另一把商户公钥) 必须拒绝。
	wrong := *env
	wrong.KeyFingerprint = "SHA256:" + strings.Repeat("a", 64)
	if _, err := OpenSensitiveResponse(priv, &wrong, aad.RequestID); !errors.Is(err, ErrKeyFingerprintMismatch) {
		t.Errorf("fingerprint mismatch: got %v, want ErrKeyFingerprintMismatch", err)
	}
}

func TestEnvelopeValidation(t *testing.T) {
	if _, err := ParseEnvelopeJSON([]byte(`{"algorithm":"RSA-OAEP-AES-128-GCM","encryptedKey":"a","ciphertext":"b","aad":"c"}`)); !errors.Is(err, ErrEnvelopeInvalid) {
		t.Error("unknown algorithm must be rejected")
	}
	if _, err := ParseEnvelopeJSON([]byte(`{"algorithm":"RSA-OAEP-AES-256-GCM","ciphertext":"b","aad":"c"}`)); !errors.Is(err, ErrEnvelopeInvalid) {
		t.Error("missing encryptedKey must be rejected")
	}
	if _, err := ParseAAD("a|b|c"); !errors.Is(err, ErrEnvelopeInvalid) {
		t.Error("aad with 3 components must be rejected")
	}
	aad, err := ParseAAD("req_1||1755600011000|SHA256:ff")
	if err != nil {
		t.Fatalf("ParseAAD: %v", err)
	}
	if aad.RouteTemplate != "" || aad.Timestamp != "1755600011000" {
		t.Fatalf("ParseAAD = %+v", aad)
	}
}

File: node/test/envelope.test.ts
import { describe, expect, it } from 'vitest';
import {
  EnvelopeCodec,
  PlutusEnvelopeError,
  buildAad,
  keyFingerprint,
  loadPrivateKey,
  parseSensitiveResponseAad,
} from '../src/index.js';
import { key, vectors } from './vectors.js';

describe('encryptedEnvelope 向量组', () => {
  it('向量数量为 3', () => {
    expect(vectors.vectors.encryptedEnvelope).toHaveLength(3);
  });

  for (const vector of vectors.vectors.encryptedEnvelope) {
    describe(`${vector.id}: ${vector.description}`, () => {
      const pair = key(vector.decryptionKey);

      it('重建的 AAD 与向量的 aadString / aadBase64 / 信封回显逐字节一致', () => {
        const aad = buildAad(vector.aadComponents);
        expect(aad.toString('utf8')).toBe(vector.aadString);
        expect(aad.toString('base64')).toBe(vector.aadBase64);
        expect(vector.envelope.aad).toBe(vector.aadBase64);
      });

      it('信封 keyFingerprint 等于解密私钥对应的公钥指纹', () => {
        expect(vector.envelope.keyFingerprint).toBe(pair.fingerprint);
        expect(vector.aadComponents.keyId).toBe(pair.fingerprint);
      });

      it('API 链信封含 encryptedPayload 兼容字段且等于 ciphertext', () => {
        expect(vector.envelope.encryptedPayload).toBe(vector.envelope.ciphertext);
      });

      it('用向量指定的私钥解密并比对明文', () => {
        const plaintext = EnvelopeCodec.openText(vector.envelope, {
          privateKey: pair.privateKeyPem,
          aad: vector.aadComponents,
        });
        expect(plaintext).toBe(vector.expectedPlaintext);
      });
    });
  }

  it('en-01 的 routeTemplate 是端点外部路径', () => {
    const v = vectors.vectors.encryptedEnvelope.find((x) => x.id === 'en-01-encrypted-request');
    expect(v?.aadComponents.routeTemplate).toBe('/card-products/10010106/shared/cards/create');
    expect(v?.decryptionKey).toBe('platform_enc');
  });

  it('敏感响应的 routeTemplate 固定为空串,AAD 中出现连续两个竖线', () => {
    for (const v of vectors.vectors.encryptedEnvelope.filter((x) => x.direction === 'sensitive_response')) {
      expect(v.aadComponents.routeTemplate).toBe('');
      expect(v.aadString).toContain('||');
      expect(v.decryptionKey).toBe('merchant_enc');
    }
  });
});

describe('encryptedEnvelope 负向', () => {
  const vector = vectors.vectors.encryptedEnvelope[1] as NonNullable<
    (typeof vectors.vectors.encryptedEnvelope)[1]
  >;
  const pair = key(vector.decryptionKey);

  it('篡改 AAD 任一分量都必须解密失败', () => {
    for (const field of ['requestId', 'routeTemplate', 'timestamp', 'keyId'] as const) {
      const aad = { ...vector.aadComponents, [field]: `${vector.aadComponents[field]}x` };
      expect(() =>
        EnvelopeCodec.open(vector.envelope, { privateKey: pair.privateKeyPem, aad }),
      ).toThrowError(PlutusEnvelopeError);
    }
  });

  it('直接使用信封回显的 aad 而不重建时,回显被篡改也应被拒绝', () => {
    const tampered = {
      ...vector.envelope,
      aad: Buffer.from(`${vector.aadString}x`, 'utf8').toString('base64'),
    };
    expect(() =>
      EnvelopeCodec.open(tampered, { privateKey: pair.privateKeyPem, aad: vector.aadComponents }),
    ).toThrowError(/rebuilt AAD does not match/);
  });

  it('篡改 ciphertext 触发 GCM 认证失败', () => {
    const blob = Buffer.from(vector.envelope.ciphertext, 'base64');
    blob[blob.length - 1] = (blob[blob.length - 1] as number) ^ 0xff;
    const tampered = { ...vector.envelope, ciphertext: blob.toString('base64') };
    expect(() =>
      EnvelopeCodec.open(tampered, { privateKey: pair.privateKeyPem, aad: vector.aadComponents }),
    ).toThrowError(/AES-GCM authentication failed/);
  });

  it('用错误的私钥解密失败', () => {
    expect(() =>
      EnvelopeCodec.open(vector.envelope, {
        privateKey: key('platform_enc').privateKeyPem,
        aad: vector.aadComponents,
      }),
    ).toThrowError(PlutusEnvelopeError);
  });

  it('算法标识不符时拒绝', () => {
    expect(() =>
      EnvelopeCodec.open(
        { ...vector.envelope, algorithm: 'RSA-OAEP-AES-128-GCM' },
        { privateKey: pair.privateKeyPem, aad: vector.aadComponents },
      ),
    ).toThrowError(/unsupported envelope algorithm/);
  });

  it('明文长度上限被强制执行', () => {
    expect(() =>
      EnvelopeCodec.open(vector.envelope, {
        privateKey: pair.privateKeyPem,
        aad: vector.aadComponents,
        maxPlaintextBytes: 4,
      }),
    ).toThrowError(/exceeds the 4 byte limit/);
  });
});

describe('敏感响应解密:从信封 aad 重建时间戳', () => {
  for (const vector of vectors.vectors.encryptedEnvelope.filter((v) => v.direction === 'sensitive_response')) {
    it(`${vector.id} 解析 AAD 分量并解密`, () => {
      const parsed = parseSensitiveResponseAad(vector.envelope.aad);
      expect(parsed).toEqual(vector.aadComponents);

      const { plaintext, aad } = EnvelopeCodec.openSensitiveResponse(vector.envelope, {
        merchantEncPrivateKey: key('merchant_enc').privateKeyPem,
        requestId: vector.aadComponents.requestId,
      });
      expect(plaintext.toString('utf8')).toBe(vector.expectedPlaintext);
      expect(aad.timestamp).toBe(vector.aadComponents.timestamp);
    });
  }

  it('关联请求 ID 不符时拒绝', () => {
    const vector = vectors.vectors.encryptedEnvelope.find((v) => v.id === 'en-02-sensitive-response');
    expect(() =>
      EnvelopeCodec.openSensitiveResponse((vector as NonNullable<typeof vector>).envelope, {
        merchantEncPrivateKey: key('merchant_enc').privateKeyPem,
        requestId: 'req_someone_elses_request',
      }),
    ).toThrowError(/requestId does not match/);
  });

  it('requestId 为空串的向量可在不传 requestId 时解密', () => {
    const vector = vectors.vectors.encryptedEnvelope.find(
      (v) => v.id === 'en-03-sensitive-response-empty-request-id',
    ) as NonNullable<(typeof vectors.vectors.encryptedEnvelope)[0]>;
    const { plaintext } = EnvelopeCodec.openSensitiveResponse(vector.envelope, {
      merchantEncPrivateKey: key('merchant_enc').privateKeyPem,
    });
    expect(plaintext.toString('utf8')).toBe(vector.expectedPlaintext);
  });
});

describe('加密方向 round-trip', () => {
  const platformEnc = key('platform_enc');
  const merchantEnc = key('merchant_enc');

  it('加密请求:sealRequest 产出的信封可由平台加密私钥解开', () => {
    const plaintext = '{"platformCardProductBizId":"pcp_example_001","quantity":2}';
    const { envelope, bodyBytes, keyId } = EnvelopeCodec.sealRequest(plaintext, {
      requestId: 'req_roundtrip_0001',
      routeTemplate: '/card-products/10010106/shared/cards/create',
      timestamp: '1755600010000',
      platformEncPublicKey: platformEnc.publicKeyPem,
    });

    expect(keyId).toBe(platformEnc.fingerprint);
    expect(envelope.algorithm).toBe('RSA-OAEP-AES-256-GCM');
    expect(envelope.encryptedPayload).toBe(envelope.ciphertext);
    expect(JSON.parse(bodyBytes.toString('utf8'))).toEqual(envelope);
    // IV(12) + 明文 + 标签(16)
    expect(Buffer.from(envelope.ciphertext, 'base64').byteLength).toBe(12 + plaintext.length + 16);

    const decrypted = EnvelopeCodec.openText(envelope, {
      privateKey: platformEnc.privateKeyPem,
      aad: {
        requestId: 'req_roundtrip_0001',
        routeTemplate: '/card-products/10010106/shared/cards/create',
        timestamp: '1755600010000',
        keyId,
      },
    });
    expect(decrypted).toBe(plaintext);
  });

  it('每次加密使用不同的一次性密钥与 IV', () => {
    const seal = (): string =>
      EnvelopeCodec.seal('{"a":1}', {
        recipientPublicKey: platformEnc.publicKeyPem,
        aad: { requestId: 'r', routeTemplate: '/p', timestamp: '1', keyId: platformEnc.fingerprint },
      }).ciphertext;
    expect(seal()).not.toBe(seal());
  });

  it('敏感响应方向:平台用商户加密公钥封装,商户私钥解开', () => {
    const plaintext = '{"cardNumber":"4111111111111111"}';
    const aad = {
      requestId: 'req_roundtrip_0002',
      routeTemplate: '',
      timestamp: '1755600011000',
      keyId: merchantEnc.fingerprint,
    };
    const envelope = EnvelopeCodec.seal(plaintext, {
      recipientPublicKey: merchantEnc.publicKeyPem,
      aad,
    });
    const { plaintext: decrypted } = EnvelopeCodec.openSensitiveResponse(envelope, {
      merchantEncPrivateKey: merchantEnc.privateKeyPem,
      requestId: aad.requestId,
    });
    expect(decrypted.toString('utf8')).toBe(plaintext);
  });

  it('Webhook 形状:含 envelopeVersion 且不含 encryptedPayload', () => {
    const envelope = EnvelopeCodec.seal('{"eventId":"x"}', {
      recipientPublicKey: merchantEnc.publicKeyPem,
      aad: { requestId: 'whd_1', routeTemplate: 'webhook', timestamp: '1', keyId: 'apk_vector_0001' },
      envelopeVersion: 1,
      includeEncryptedPayload: false,
    });
    expect(envelope.envelopeVersion).toBe(1);
    expect('encryptedPayload' in envelope).toBe(false);
    expect(Object.keys(envelope).sort()).toEqual(
      ['aad', 'algorithm', 'ciphertext', 'encryptedKey', 'envelopeVersion', 'keyFingerprint'].sort(),
    );
  });

  it('信封 keyFingerprint 与接收方公钥指纹一致', () => {
    const envelope = EnvelopeCodec.seal('x', {
      recipientPublicKey: merchantEnc.publicKeyPem,
      aad: { requestId: '', routeTemplate: '', timestamp: '1', keyId: merchantEnc.fingerprint },
    });
    expect(envelope.keyFingerprint).toBe(keyFingerprint(loadPrivateKey(merchantEnc.privateKeyPem)));
  });
});
